src/context.rs

```diff
@@ -189,82 +189,57 @@ impl Stream {
-static mut _HOST_ALLOCATOR: Option<StaticHostAllocator> = None;
-static mut _SMALL_HOST_ALLOCATOR: Option<SmallStaticHostAllocator> = None;
-
 pub(crate) fn _alloc() -> &'static StaticDeviceAllocator {
-    unsafe {
-        _DEVICE_ALLOCATOR
-            .as_ref()
-            .expect("device allocator should be initialized")
-    }
+    &get_context().device_allocator
 }
 
 pub(crate) fn _small_alloc() -> &'static SmallStaticDeviceAllocator {
-    unsafe {
-        _SMALL_DEVICE_ALLOCATOR
-            .as_ref()
-            .expect("small device allocator should be initialized")
-    }
+    &get_context().small_device_allocator
 }
 pub(crate) fn _host_alloc() -> &'static StaticHostAllocator {
-    unsafe {
-        _HOST_ALLOCATOR
-            .as_ref()
-            .expect("host allocator should be initialized")
-    }
+    &get_context().host_allocator
 }
 
 pub(crate) fn _small_host_alloc() -> &'static SmallStaticHostAllocator {
-    unsafe {
-        _SMALL_HOST_ALLOCATOR
-            .as_ref()
-            .expect("small host allocator should be initialized")
-    }
+    &get_context().small_host_allocator
 }
 
-static mut _SETUP_CACHE: Option<SetupCache> = None;
-
 pub(crate) fn _setup_cache_get() -> Option<&'static mut SetupCache> {
-    unsafe { _SETUP_CACHE.as_mut() }
+    get_context_mut().setup_cache.as_mut()
 }
 
 pub(crate) fn _setup_cache_set(value: SetupCache) {
-    unsafe {
-        assert!(_SETUP_CACHE.is_none());
-        _SETUP_CACHE = Some(value)
-    }
+    assert!(_setup_cache_get().is_none());
+    get_context_mut().setup_cache = Some(value);
 }
 
 pub(crate) fn _setup_cache_reset() {
-    unsafe { _SETUP_CACHE = None }
+    get_context_mut().setup_cache = None;
 }
 
-static mut _STRATEGY_CACHE: Option<HashMap<Vec<[F; 4]>, CacheStrategy>> = None;
-
 pub(crate) fn _strategy_cache_get() -> &'static mut HashMap<Vec<[F; 4]>, CacheStrategy> {
-    unsafe {
-        _STRATEGY_CACHE
-            .as_mut()
-            .expect("strategy cache should be initialized")
-    }
+    &mut get_context_mut().strategy_cache
 }
 pub(crate) fn _strategy_cache_reset() {
-    unsafe { _STRATEGY_CACHE = Some(HashMap::new()) }
+    get_context_mut().strategy_cache.clear();
 }
 
 pub(crate) fn is_prover_context_initialized() -> bool {
-    unsafe {
-        _CUDA_CONTEXT.is_some()
-            & _EXEC_STREAM.is_some()
-            & _H2D_STREAM.is_some()
-            & _D2H_STREAM.is_some()
-            & _DEVICE_ALLOCATOR.is_some()
-            & _SMALL_DEVICE_ALLOCATOR.is_some()
-            & _HOST_ALLOCATOR.is_some()
-            & _SMALL_HOST_ALLOCATOR.is_some()
-            & _STRATEGY_CACHE.is_some()
-    }
+    unsafe { CONTEXT.is_some() }
+}
+
+pub(crate) fn _l2_cache_size() -> usize {
+    get_context().l2_cache_size
+}
+
+pub(crate) fn _compute_capability_major() -> u32 {
+    get_context().compute_capability_major
+}
+
+pub(crate) fn _aux_streams() -> &'static [CudaStream; NUM_AUX_STREAMS_AND_EVENTS] {
```

Comment:
Is this the expected way for external code to grab a helper stream?

Reply:
yes, `NUM_AUX_STREAMS_AND_EVENTS` number of helper events and streams is permanently kept available